ompi/mca/op/aarch64/op_aarch64_functions.c

```diff
@@ -0,0 +1,641 @@
+/*
+ * Copyright (c) 2019      The University of Tennessee and The University
+ *                         of Tennessee Research Foundation.  All rights
+ *                         reserved.
+ * Copyright (c) 2019      Arm Ltd.  All rights reserved.
+ * Copyright (c) 2024      NVIDIA Corporation.  All rights reserved.
+ *
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+
+#ifdef HAVE_SYS_TYPES_H
+#include <sys/types.h>
+#endif
+#include "opal/util/output.h"
+
+#include "ompi/op/op.h"
+#include "ompi/mca/op/op.h"
+#include "ompi/mca/op/base/base.h"
+#include "ompi/mca/op/aarch64/op_aarch64.h"
+
+#if defined(GENERATE_SVE_CODE)
+#    include <arm_sve.h>
+#define OMPI_OP_TYPE_PREPEND sv
+#define OMPI_OP_OP_PREPEND sv
+#define APPEND   _sve
+#elif defined(GENERATE_NEON_CODE)
+#    include <arm_neon.h>
+#define OMPI_OP_TYPE_PREPEND
+#define OMPI_OP_OP_PREPEND v
+#define APPEND   _neon
+#else
+#error we should not reach this
+#endif /* OMPI_MCA_OP_HAVE_SVE */
+
+/*
+ * Concatenate preprocessor tokens A and B without expanding macro definitions
+ * (however, if invoked from a macro, macro arguments are expanded).
+ */
+#define OP_CONCAT_NX(A, B) A ## B
+
+/*
+ * Concatenate preprocessor tokens A and B after macro-expanding them.
+ */
+#define OP_CONCAT(A, B) OP_CONCAT_NX(A, B)
+
+#if defined(GENERATE_SVE_CODE)
+#    define svcnt(X)           \
+        _Generic((X),          \
+            int8_t: svcntb,    \
+            uint8_t: svcntb,   \
+            int16_t: svcnth,   \
+            uint16_t: svcnth,  \
+            int32_t: svcntw,   \
+            uint32_t: svcntw,  \
+            int64_t: svcntd,   \
+            uint64_t: svcntd,  \
+            float32_t: svcntw, \
+            float64_t: svcntd)()
+#else
+#define DUMP2(out, in1, in2) \
+        case 2: (out)[1] = current_func((in1)[1], (in2)[1]); \
+        case 1: (out)[0] = current_func((in1)[0], (in2)[0]);
+#define DUMP4(out, in1, in2) \
+        case 4: (out)[3] = current_func((in1)[3], (in2)[3]); \
+        case 3: (out)[2] = current_func((in1)[2], (in2)[2]); \
+        DUMP2(out, in1, in2)
+#define DUMP8(out, in1, in2) \
+        case 8: (out)[7] = current_func((in1)[7], (in2)[7]); \
+        case 7: (out)[6] = current_func((in1)[6], (in2)[6]); \
+        case 6: (out)[5] = current_func((in1)[5], (in2)[5]); \
+        case 5: (out)[4] = current_func((in1)[4], (in2)[4]); \
+        DUMP4(out, in1, in2)
+
+#    define neon_loop(out, in1, in2)                \
+_Generic((*(out)), \
+ int8_t:    __extension__({ switch (how_much) { DUMP8(out, in1, in2) }}),\
+ uint8_t:   __extension__({ switch (how_much) { DUMP8(out, in1, in2) }}),\
+ int16_t:   __extension__({ switch (how_much) { DUMP8(out, in1, in2) }}), \
```

Comment:
Why are the single-byte types not getting DUMP16 here?

Reply:
Because C macros suck. This code only works for gcc and clang, by using __extension__. Otherwise I might have to move this code into a header file and include it multiple times. Ugly and impossible to debug.

If you have any better ideas please share.